Introduce Parameter Object: QueryIntent

matthewd opened the pull request:
This has been on my todo list for a while, as we've continued to gradually accumulate extra context that we end up needing to pass along with queries.

~~**This is a draft** -- for exploration, I've changed a number of low-level-but-public APIs. Even if we agree that's a reasonable end-goal, they would require deprecations along the way.~~

Affected Public APIs:

`exec_update` and `exec_delete` are deprecated, pointing callers to `update` and `delete` instead. Those methods (the only internal callers) now just `affected_rows(raw_execute(intent))` directly.

`exec_insert` is likewise deprecated in favour of `insert`. We [currently?] still need to keep it for overriding, so the internal intent-using one is now `_exec_insert`. That's ugly, but I'm hopeful we can later evolve it away, and failing that we'll be able to reclaim `exec_insert` post-deprecation.

`log` is now called exclusively with an `intent`, but still accepts `sql, name, binds` calls too (with deprecation). In practice I don't anticipate many people calling it with existing parameters, but it is technically documented (as a `:doc:` private method). I anticipate most external users of `log` are overriding/wrapping it, so will be broken by the signature change on our existing calls -- but I think that's unavoidable.

matthewd committed:
Recover if a test interruption has left the old_references table around

matthewd committed:
Introduce QueryIntent

This is an adapter-internal object that represents a caller's desire
that a given query be run, encapsulating the SQL, binds, and various
control parameters like allow_retry.

Immutable here, but mutable attributes to follow.

matthewd committed:
Mutate QueryIntent as it works its way through execution

Carry the type-casted binds and notification payload once they're
available; update #async to reflect whether the query was actually run
asynchronously.

matthewd committed:
Move SQL preprocessing into QueryIntent

The work stays in the adapter, but we now construct the intent with the
original SQL, to be stored separately from the final preprocessed form.

matthewd committed:
Move Arel preprocessing into QueryIntent too

Following the SQL preprocessing, further allow QueryIntent to be
constructed with an Arel-shaped query.

matthewd committed:
Deprecate exec_{insert,update,delete}

These are purely less-powerful versions of their non-prefixed public API
equivalents, and the adapter has far too many query-running methods to
justify keeping obscure variants for compatibility.

For update and delete, these are now entirely vestigial. For insert,
some adapters need to be able to intervene, so I've moved the real work
to _exec_insert.

byroot reviewed (commented):
I had something like that in mind for a while, but I really think it should just be a "query object".

matthewd commented:
Mostly just for my own future reference / comparison, adapter methods that accept and run a query:

Before:

<img width="2119" height="1438" alt="mermaid-diagram-2025-10-29-170613" src="https://github.com/user-attachments/assets/d6871e16-624a-43d4-99a6-beee38117cef" />

Now:

<img width="2119" height="1438" alt="mermaid-diagram-2025-10-29-175554" src="https://github.com/user-attachments/assets/1b86ea09-2dcc-4447-b1dd-023bed7fd60e" />

(Blue = public API, purple = nodoc, orange = private, blue-grey = deprecated, grey = adapter-external [calling non-API method], red = internal core; black arrow = passing Arel/SQL query, pink arrow = passing intent)

byroot commented on activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -1197,23 +1197,53 @@ def translate_exception_class(native_error, sql, binds)
+            raise ex.set_query(intent.processed_sql, intent.binds)
+          else
+            raise ex.set_query(sql, binds)
+          end
```
Nitpick: I'd move that rescue in each branch of the `if`, so we have a single conditional instead of two.

byroot commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -51,6 +51,15 @@ def to_sql_and_binds(arel_or_sql_string, binds = [], preparable = nil, allow_ret
+        sql, binds, preparable, allow_retry = to_sql_and_binds(intent.arel, intent.binds, intent.prepare, intent.allow_retry)
+        intent.set_compiled_result(raw_sql: sql, binds: binds, prepare: preparable, allow_retry: allow_retry)
+      end
+      private :compile_arel_in_intent
```
I believe this logic should move inside `QueryAttempt`, e.g. `intent.compile(self)`. It would be better encapsulation.

byroot commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -158,21 +169,40 @@ def exec_query(sql, name = "SQL", binds = [], prepare: false)
       # the executed +sql+ statement.
       def exec_delete(sql, name = nil, binds = [])
+        ActiveRecord.deprecator.warn("exec_delete is deprecated and will be removed in Rails 8.2. " \
+                                     "Please use #delete instead.")
```
Nitpick: You could use `deprecate_methods` helper for all of these.

byroot reviewed (approved):
It's big so I haven't reviewed everything in close detail, but seems like a positive refactor to me, we can always refine later.

matthewd committed:
Pass QueryIntent to #log

It was documented despite being private, so we continue accepting the
old separated parameter set with a deprecation warning.

The method is no longer documented: since the consolidation onto
perform_query, individual adapters should not need to know about it.

matthewd committed:
Pass QueryIntent to #perform_query

This moves intent-unpacking to the same boundary where we start talking
to the raw connection.

bquorning commented on activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:
```diff
@@ -564,18 +598,24 @@ def default_insert_value(column) # :nodoc:
-              handle_warnings(result, sql)
+        def raw_execute(intent)
+          # Handle Arel compilation if needed
+          compile_arel_in_intent(intent)
```
I may be wrong, but is this an unintended API change? I think that `#raw_execute` did not previously accept Arel, but only raw sql? And it looks like you changed all the internal code to compile Arel into sql before calling `#raw_execute`.
